[Training] Add `datasets` version of LCM LoRA SDXL

sayakpaul opened the pull request:
# What does this PR do?

Add a `datasets` compatible variant of https://github.com/huggingface/diffusers/blob/main/examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py. It also adapts a couple of best practices from `peft`:

* Instead of initializing two UNets (teacher and student), it leverages the `disable_adapters()` and `enable_adapters()` functions. In this case, the teacher is without any adapters and the student is with LoRA. We only update the LoRA params in the student. 
* Makes use of `peft` utility modules positioning it as a utility library rather than a modelling library. 

Running a couple experiments. Will report back the findings. But should be more or less ready to be reviewed. 

## Basic training command

```bash
export MODEL_NAME="stabilityai/stable-diffusion-xl-base-1.0"
export OUTPUT_DIR="lora-lora-sdxl"
export DATASET_NAME="lambdalabs/pokemon-blip-captions"
export VAE_PATH="madebyollin/sdxl-vae-fp16-fix"

accelerate launch train_lcm_distill_lora_sdxl.py \
  --pretrained_teacher_model=${MODEL_NAME}  \
  --pretrained_vae_model_name_or_path=${VAE_PATH} \
  --output_dir=${OUTPUT_DIR} \
  --mixed_precision="fp16" \
  --dataset_name=$DATASET_NAME \
  --resolution=1024 \
  --train_batch_size=4 \
  --gradient_accumulation_steps=1 \
  --gradient_checkpointing \
  --use_8bit_adam \
  --learning_rate=1e-4 --loss_type="huber" \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=1500 \
  --checkpointing_steps=50 \
  --validation_steps=25 \
  --seed="0" \
  --report_to="wandb" \
  --push_to_hub
```

## TODOs

- [x] Update the PR with results
- [ ] Docs
- [x] Tests

pcuenca reviewed (commented):
Looks good so far, let me know if you need help testing / debugging

sayakpaul commented:
It works, needs hyperparameter tuning. If you could take it to a test run, that would be great!

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1356 @@
+    parser.add_argument(
+        "--learning_rate",
+        type=float,
+        default=1e-4,
```
Since that's the default we use.
```suggestion
        default=1e-6,
```

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1356 @@
+    parser.add_argument("--adam_epsilon", type=float, default=1e-08, help="Epsilon value for the Adam optimizer")
+    parser.add_argument("--max_grad_norm", default=1.0, type=float, help="Max gradient norm.")
+    # ----Diffusion Training Arguments----
+    parser.add_argument(
+        "--proportion_empty_prompts",
+        type=float,
+        default=0.0,
+        help="Proportion of image prompts to be replaced with empty strings. Defaults to 0 (no prompt replacement).",
+    )
```
Maybe let's remove this. Not really required here.

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1356 @@
+    # 8. Add LoRA to the student U-Net, only the LoRA projection matrix will be updated by the optimizer.
+    lora_config = LoraConfig(
+        r=args.lora_rank,
+        target_modules=[
+            "to_q",
+            "to_k",
+            "to_v",
+            "to_out.0",
+            "proj_in",
+            "proj_out",
+            "ff.net.0.proj",
+            "ff.net.2",
+            "conv1",
+            "conv2",
+            "conv_shortcut",
+            "downsamplers.0.conv",
+            "upsamplers.0.conv",
+            "time_emb_proj",
```
for later:

We could also think of making this an argument.

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1356 @@
+            if accelerator.is_main_process:
+                unet_ = accelerator.unwrap_model(unet)
+                lora_state_dict = get_peft_model_state_dict(unet_, adapter_name="default")
+                StableDiffusionXLPipeline.save_lora_weights(os.path.join(output_dir, "unet_lora"), lora_state_dict)
```
Let's make sure we can load these saved weights for inference. Last time I tried there was a error in loading, but it could be because of my mistake (not fully familer with LoRA logic in diffusers).

I ended up converting the state_dict to kohya format and manually saving it with `torch.save`

sayakpaul replied:
The problem is that we're trying to serialize the `peft` state dict with `save_lora_layers()` and that shouldn't be the case because the `peft` state dict also needs to be supplemented with an adapter config file. This is what's being done in the load model hook where we make use of `load_adapter()`. 

For the time being, I have modified it so that we output the state dict in the Kohya format so that we know that it's gonna be compatible with `load_lora_weights()`. But I will revisit it so that we can output a diffusers-compatible single file checkpoint (by making use of `convert_state_dict_to_diffusers()`). 

IIRC, we cannot yet load a `peft` LoRA checkpoint directly with `load_lora_weights()`. This should be fixed in a separate PR if that's the case.

Cc: @pacman100 @younesbelkada

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1356 @@
+        revision=args.teacher_revision,
+    )
+
+    # 5. Load teacher U-Net from SD-XL checkpoint
+    teacher_unet = UNet2DConditionModel.from_pretrained(
+        args.pretrained_teacher_model, subfolder="unet", revision=args.teacher_revision
+    )
```
For later:

We could potentially avoid loading two unets. Since we are doing LoRA, we could disable the LoRA when getting teacher prediction.

sayakpaul replied:
@pacman100 we're using `peft` here. IIRC, there's a nice context manager in `peft` that lets us run a model without the adapters. Could you provide a reference?

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1356 @@
+
+                # encode pixel values with batch size of at most 8
+                latents = []
+                for i in range(0, pixel_values.shape[0], 8):
```
Could also make this an argument.

sayakpaul replied:
Done in 6b2e42f2

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1356 @@
+                # noisy_latents with both the conditioning embedding c and unconditional embedding 0
+                # Get teacher model prediction on noisy_latents and conditional embedding
+                with torch.no_grad():
+                    with torch.autocast("cuda"):
```
```suggestion
                    with torch.autocast("cuda", dtype=weight_dtype):
```

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1356 @@
+
+                # Get target LCM prediction on x_prev, w, c, t_n
+                with torch.no_grad():
+                    with torch.autocast("cuda", enabled=True, dtype=weight_dtype):
```
```suggestion
                    with torch.autocast("cuda", dtype=weight_dtype):
```

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1356 @@
+                        target_noise_pred = unet(
+                            x_prev.float(),
+                            timesteps,
+                            timestep_cond=None,
```
```suggestion
```

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1356 @@
+                noise_pred = unet(
+                    noisy_model_input,
+                    start_timesteps,
+                    timestep_cond=None,
```
```suggestion
```

patil-suraj reviewed (approved):
Thanks a lot for working on this. Just left some comments. Most importantly:

Let's make sure we can load the saved lora with the pipeline for inference. I had some issues with it when I tried. 

Everything else looks good !

sayakpaul commented:
> Let's make sure we can load the saved lora with the pipeline for inference. I had some issues with it when I tried.

Could you post the snippet with which you tried and the error you got?

patil-suraj commented:
> > Let's make sure we can load the saved lora with the pipeline for inference. I had some issues with it when I tried.
> 
> Could you post the snippet with which you tried and the error you got?

When I try to load the lora saved with script using 

```py
pipe.load_lora_weights(path_to_saved_lora, weight_name="pytorch_lora_weights.safetensors")
```

I get 
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In [13], line 1
----> 1 pipe.load_lora_weights("valhalla/lcm-sdxl-distill-lora-k5", weight_name="pytorch_lora_weights.safetensors")

File ~/diffusers/src/diffusers/loaders.py:3233, in StableDiffusionXLLoraLoaderMixin.load_lora_weights(self, pretrained_model_name_or_path_or_dict, adapter_name, **kwargs)
   3230 if not is_correct_format:
   3231     raise ValueError("Invalid LoRA checkpoint.")
-> 3233 self.load_lora_into_unet(
   3234     state_dict, network_alphas=network_alphas, unet=self.unet, adapter_name=adapter_name, _pipeline=self
   3235 )
   3236 text_encoder_state_dict = {k: v for k, v in state_dict.items() if "text_encoder." in k}
   3237 if len(text_encoder_state_dict) > 0:

File ~/diffusers/src/diffusers/loaders.py:1630, in LoraLoaderMixin.load_lora_into_unet(cls, state_dict, network_alphas, unet, low_cpu_mem_usage, adapter_name, _pipeline)
   1627     if "lora_B" in key:
   1628         rank[key] = val.shape[1]
-> 1630 lora_config_kwargs = get_peft_kwargs(rank, network_alphas, state_dict, is_unet=True)
   1631 lora_config = LoraConfig(**lora_config_kwargs)
   1633 # adapter_name

File ~/diffusers/src/diffusers/utils/peft_utils.py:122, in get_peft_kwargs(rank_dict, network_alpha_dict, peft_state_dict, is_unet)
    120 rank_pattern = {}
    121 alpha_pattern = {}
--> 122 r = lora_alpha = list(rank_dict.values())[0]
    124 if len(set(rank_dict.values())) > 1:
    125     # get the rank occuring the most number of times
    126     r = collections.Counter(rank_dict.values()).most_common()[0][0]
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_5778). All of your documentation changes will be reflected on that endpoint.

sayakpaul commented:
@patil-suraj here's a summary of the changes:

* No teacher UNet is being loaded separately. Thanks to `disable_adapter()` context (cc: @younesbelkada) 
* Cleaned the with contexts a bit. 
* Fixed the serialization. 

The following should work now:

```python
from diffusers import StableDiffusionXLPipeline
import torch


pipeline_id = "stabilityai/stable-diffusion-xl-base-1.0"
ckpt_id = "sayakpaul/lora-lcm-sdxl-new"

pipeline = StableDiffusionXLPipeline.from_pretrained(pipeline_id, torch_dtype=torch.float16).to("cuda")
pipeline.load_lora_weights(ckpt_id, use_auth_token=True)
```

We're now serializing the final checkpoint in the `diffusers` native format. 

Currently, conducting an experiment with the following command:

```bash
CUDA_VISIBLE_DEVICES=1 accelerate launch train_lcm_distill_lora_sdxl.py \
  --pretrained_teacher_model=${MODEL_NAME}  \
  --pretrained_vae_model_name_or_path=${VAE_PATH} \
  --output_dir="lora-lcm-sdxl-new" \
  --mixed_precision="fp16" \
  --dataset_name=$DATASET_NAME \
  --resolution=1024 \
  --train_batch_size=16 \
  --gradient_accumulation_steps=1 \
  --gradient_checkpointing \
  --use_8bit_adam \
  --learning_rate=1e-6 --loss_type="huber" \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=5000 \
  --checkpointing_steps=100 \
  --validation_steps=50 \
  --seed="0" \
  --report_to="wandb" \
  --push_to_hub
```

It'd be very nice to have your 👀 on this.

patrickvonplaten commented:
cc @patil-suraj

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1354 @@
+                    "xFormers 0.0.16 cannot be used for training in some GPUs. If you observe problems during training, please update xFormers to at least 0.0.17. See https://huggingface.co/docs/diffusers/main/en/optimization/xformers for more details."
+                )
+            unet.enable_xformers_memory_efficient_attention()
+            # teacher_unet.enable_xformers_memory_efficient_attention()
```
```suggestion
```

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1354 @@
+                # Get teacher model prediction on noisy_latents and conditional embedding
+                # Notice that we're disabling the adapter layers within the `unet` and then it becomes a
+                # regular teacher. This way, we don't have to separately initialize a teacher UNet.
+                with torch.no_grad() and torch.autocast(
+                    str(accelerator.device), dtype=weight_dtype
+                ) and unet.disable_adapter():
```
Since we are just using one `unet` now, this `autocast` might not be necessary. The `unet` is already prepared with `accelerate` which wraps the model in `autocast` when doing mixed-precision training

sayakpaul replied:
Kept it because the `unet` that's being trained is also under `autocast` (taken from the original script).

patil-suraj reviewed (commented):
Looking very good, just left some suggestions. We should remove the empty prompt embeds options here, since we train the model with CFG it's not necessary. Apart from  this, it's in a pretty good state already.

patil-suraj commented:
Curious to know how are the results with smaller datasets :)

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1354 @@
+        # create custom saving & loading hooks so that `accelerator.save_state(...)` serializes in a nice format
+        def save_model_hook(models, weights, output_dir):
+            if accelerator.is_main_process:
+                unet_ = accelerator.unwrap_model(unet)
+                # save weights in peft format to be able to load them back
+                unet_.save_pretrained(output_dir)
```
Sounds good!

sayakpaul commented:
@DN6 we can't add `pip install peft` in the PR test workflow either as that swaps the training backend to use `peft` which is not fully supported for the other training examples. 

I see the following as a potential workaround. In the workflow, we run all the example tests except for `test_text_to_image_lcm_lora_sdxl` without installing `peft`. Then we add a command to install `peft` and then just test `test_text_to_image_lcm_lora_sdxl`. Does this work for you? 

Later, when the examples are fully equipped with `peft` we can revisit it. LMK.

sayakpaul commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1319 @@
+                # regular teacher. This way, we don't have to separately initialize a teacher UNet.
+                using_cuda = "cuda" in str(accelerator.device)
+                with torch.no_grad() and torch.autocast(
+                    str(accelerator.device), dtype=weight_dtype if using_cuda else torch.bfloat16, enabled=using_cuda
```
@patil-suraj I am not super happy about how I am doing it here but it's the only way I found to make the test pass on a CPU. Let me know if you have better ideas.

younesbelkada commented:
@sayakpaul I can confirm all issues with respect to failing example CI are fixed with https://github.com/huggingface/diffusers/pull/5388. Perhaps we can merge first https://github.com/huggingface/diffusers/pull/5388 and I can work on the saving adapter config as a follow up PR. What do you think?

sayakpaul commented:
I will have a look at #5388.

younesbelkada commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1319 @@
+                # be independently loaded via `load_lora_weights()`.
+                peft_state_dict = get_peft_model_state_dict(unet, adapter_name="default")
+                diffusers_state_dict = convert_state_dict_to_diffusers(peft_state_dict)
+                diffusers_state_dict = {
+                    f"{module_name.replace('base_model.model.', '')}.{module_name}": param
+                    for module_name, param in diffusers_state_dict.items()
+                }
```
```suggestion
```
This is not needed in case you don't use the `PeftModel` interface, only `get_peft_model_state_dict(unet)` should suffice

younesbelkada commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1319 @@
+            "time_emb_proj",
+        ],
+    )
+    unet = get_peft_model(unet, lora_config)
```
you can simply do `unet.add_adapter(lora_config)` and by-pass the `PeftModel` interface which is what I am doing in https://github.com/huggingface/diffusers/pull/5388 
Although it is possible to use a `PeftModel` through `get_peft_model` I would rather encourage users to use `add_adapter` interface to have an API which is consistent across all examples.

```suggestion
    unet.add_adapter(lora_config)
```

younesbelkada commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1319 @@
+
+    # 12. Optimizer creation
+    optimizer = optimizer_class(
+        unet.parameters(),
```
```suggestion
        filter(lambda p: p.requires_grad, unet.parameters()),
```
to only select trainable parameters

younesbelkada commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1319 @@
+            if accelerator.is_main_process:
+                unet_ = accelerator.unwrap_model(unet)
+                # save weights in peft format to be able to load them back
+                unet_.save_pretrained(output_dir)
```
```suggestion
                unet_.peft_config.save_pretrained(output_dir)
```
The save_pretrained here will save both `adapter_config.json` and `adapter_model.bin`, but the `.bin` gets ignored since we don't use it when we load lora weights in PEFT + diffusers. Since the lora weights are properly saved below, I propose to only save the adapter config as such

sayakpaul replied:
@younesbelkada we're using `load_adapter()` on the UNet.

sayakpaul replied:
Ah I see what you mean. See if the current changes make sense.

younesbelkada commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1319 @@
+                using_cuda = "cuda" in str(accelerator.device)
+                with torch.no_grad() and torch.autocast(
+                    str(accelerator.device), dtype=weight_dtype if using_cuda else torch.bfloat16, enabled=using_cuda
+                ) and unet.disable_adapter():
```
```suggestion
                ):
```
In case you would go without `PeftModel` interface, you need to remove this and accept my suggestion above and below

younesbelkada commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1319 @@
+                # Notice that we're disabling the adapter layers within the `unet` and then it becomes a
+                # regular teacher. This way, we don't have to separately initialize a teacher UNet.
+                using_cuda = "cuda" in str(accelerator.device)
+                with torch.no_grad() and torch.autocast(
```
```suggestion
                unet.disable_adapters()
                with torch.no_grad() and torch.autocast(
```

younesbelkada commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1319 @@
+                        encoder_hidden_states=prompt_embeds.to(weight_dtype),
+                        added_cond_kwargs={k: v.to(weight_dtype) for k, v in encoded_text.items()},
+                    ).sample
+                    cond_pred_x0 = predicted_origin(
```
```suggestion
                    # re-enable unet adapters
                    unet.enable_adapters()
                    cond_pred_x0 = predicted_origin(
```

younesbelkada reviewed (commented):
Hi @sayakpaul 
Thanks for your great efforts!
I left few comments regarding the current state of the fine-tuning script, right now the script deviates a bit from the PEFT integration in diffusers as it utilises the `PeftModel` interface. Although it works fine, it might lead to some confusion (why `load_adapter` for inference that modifies in-place the model and `get_peft_model` that suddenly returns a `PeftModel` for training? ). IMO when using PEFT integration (in diffusers and/or transfomers) we should see peft as an utility library rather than a modeling library
I propose to slightly refactor the training logic, which should also work as expected, following https://github.com/huggingface/diffusers/pull/5388 - If that's something we want to ship asap that's totally fine and can be re-worked in a follow up PR, let me know if you need any help!

sayakpaul commented:
@younesbelkada I resolved all your comments. Could you take a look again? Thanks so much in advance!

sayakpaul commented:
Tests will pass after https://github.com/huggingface/diffusers/pull/5388 is merged.

sayakpaul commented:
Also, @patil-suraj, I launched an experiment yesterday: 

```bash
export MODEL_NAME="stabilityai/stable-diffusion-xl-base-1.0"
export OUTPUT_DIR="lora-lora-sdxl-new"
export DATASET_NAME="lambdalabs/pokemon-blip-captions"
export VAE_PATH="madebyollin/sdxl-vae-fp16-fix"

CUDA_VISIBLE_DEVICES=1 accelerate launch train_lcm_distill_lora_sdxl.py \
  --pretrained_teacher_model=${MODEL_NAME}  \
  --pretrained_vae_model_name_or_path=${VAE_PATH} \
  --output_dir="lora-lcm-sdxl-new" \
  --mixed_precision="fp16" \
  --dataset_name=$DATASET_NAME \
  --resolution=1024 \
  --train_batch_size=16 \
  --gradient_accumulation_steps=1 \
  --gradient_checkpointing \
  --use_8bit_adam \
  --learning_rate=1e-5 --loss_type="huber" \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=5000 \
  --checkpointing_steps=100 \
  --validation_steps=50 \
  --seed="0" \
  --report_to="wandb" \
  --push_to_hub
```

The results (https://wandb.ai/sayakpaul/text2image-fine-tune/runs/ahqu0wvp) are random noise. Anything crucial I am missing? Any usual suspect?

pcuenca reviewed (commented):
A couple of things I saw while testing:

- I ran out of GPU memory while running the _second_ evaluation. I verified my max batch size by running the script using `--validation_steps=5` to verify that I could train, evaluate and go back to training successfully. But when I launched the training run for real with evaluation and checkpointing every 100 steps, it crashed at the 200 step mark. I'm not sure what could be the reason for this. I decreased my batch size (slightly, 12 to 10) after that, and the same thing happened again 
- There are many images smaller than 1024 pixels in the pokemon-blip-captions dataset (didn't count how many). Would that be a problem for SDXL training?

sayakpaul commented:
> I ran out of GPU memory while running the second evaluation. I verified my max batch size by running the script using --validation_steps=5 to verify that I could train, evaluate and go back to training successfully. But when I launched the training run for real with evaluation and checkpointing every 100 steps, it crashed at the 200 step mark. I'm not sure what could be the reason for this. I decreased my batch size (slightly, 12 to 10) after that, and the same thing happened again

Didn't see this with the following (I am using a single GPU from the DGX): 

```bash
export MODEL_NAME="stabilityai/stable-diffusion-xl-base-1.0"
export OUTPUT_DIR="lora-lora-sdxl-new"
export DATASET_NAME="lambdalabs/pokemon-blip-captions"
export VAE_PATH="madebyollin/sdxl-vae-fp16-fix"

CUDA_VISIBLE_DEVICES=1 accelerate launch train_lcm_distill_lora_sdxl.py \
  --pretrained_teacher_model=${MODEL_NAME}  \
  --pretrained_vae_model_name_or_path=${VAE_PATH} \
  --output_dir="lora-lcm-sdxl-new" \
  --mixed_precision="fp16" \
  --dataset_name=$DATASET_NAME \
  --resolution=1024 \
  --train_batch_size=24 \
  --gradient_accumulation_steps=1 \
  --gradient_checkpointing \
  --use_8bit_adam \
  --lora_rank=16 \
  --learning_rate=1e-5 --loss_type="huber" --adam_weight_decay=0.0 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=5000 \
  --checkpointing_steps=100 \
  --validation_steps=50 \
  --seed="0" \
  --report_to="wandb" \
  --push_to_hub
```

> There are many images smaller than 1024 pixels in the pokemon-blip-captions dataset (didn't count how many). Would that be a problem for SDXL training?

Could be but SDXL doesn't really excel at lower resolutions than that from what I have seen. Currently, I don't have a good workaround other than suggesting the use of an upscaler to upsample the images to 1024x1024.

younesbelkada commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1324 @@
+                    x_prev = solver.ddim_step(pred_x0, pred_noise, index)
+
+                # re-enable unet adapters
+                unet.enable_adapters()
```
```suggestion
                unet.enable_lora()
```

younesbelkada commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1324 @@
+                # Notice that we're disabling the adapter layers within the `unet` and then it becomes a
+                # regular teacher. This way, we don't have to separately initialize a teacher UNet.
+                using_cuda = "cuda" in str(accelerator.device)
+                unet.disable_adapters()
```
I just realized we do have a duplicated method `disable_lora` / `enable_lora` here: https://github.com/huggingface/diffusers/blob/main/src/diffusers/loaders/unet.py#L536 let's just keep that in mind and only use `disable_adapters` in the future

younesbelkada reviewed (approved):
Impressive work @sayakpaul ! All good on PEFT side!

ArturFormella commented:
What about other files with the same problem?
- train_lcm_distill_lora_sd_wds.py
- train_lcm_distill_sd_wds.py
- train_lcm_distill_sdxl_wds.py

More info:
https://github.com/huan085128/lcm_lora/issues/1
https://github.com/huggingface/diffusers/blob/main/examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py#L506C15-L506C15

sayakpaul commented:
@ArturFormella this PR should provide a good reference for anyone willing to adjust the scripts accordingly. Plus @pcuenca is working on #5908.

sayakpaul commented:
@BenjaminBossan pinging you since Sourab is busy and Younes is OOO. 

In this PR, I am doing what I described here: https://github.com/huggingface/diffusers/pull/5778#issue-1989907771. The training, however, seems to be completely off as it's yielding just noise: https://wandb.ai/sayakpaul/text2image-fine-tune/runs/ax6w9q0x?workspace=user-sayakpaul. I don't think it's just a matter of hyperparameters. 

I suspect some `peft` related changes that I did in this PR might have caused it. I am unable to see any potential suspects in comparison to what we have in https://github.com/huggingface/diffusers/blob/main/examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py (don't worry about the dataloading code). Would you be able to take a look and advise regarding the `peft` related changes?

BenjaminBossan commented:
Hi @sayakpaul I took a look at the script and also compared it to the other one you linked. I have no experience in training diffusion models, so I cannot judge if the generated output could be solved by some better hyper-parameters or if something else is completely off. The diff between the two scripts is quite large, so I could not pinpoint what difference causes the trouble.

The main difference that involves PEFT seems to be the way that the teacher network is being used via `disable_adapters()`. IIUC, the teacher net is completely frozen, whereas the student net is teacher + LoRA parameters on top, the latter being updated. Maybe something worth checking would be if `requires_grad` is set correctly on all params after `disable_adapters()` is called, and whether it is re-set correctly after `enable_adapters()` is being called. The diffusers implementation of these methods looks correct to me, but if training issues are PEFT related, this is what I would check first.

sayakpaul commented:
Thanks @BenjaminBossan for your inputs. I will get to them and let you know.

sayakpaul commented:
@BenjaminBossan sorry for the delay here. Big log file here: https://huggingface.co/datasets/sayakpaul/sample-datasets/blob/main/logs.txt. 

Some findings: 

* I started by printing some info on the original `params_to_optimize` here: https://github.com/huggingface/diffusers/blob/8fecdda23cefc30053a511db597b9c017b8fcaa1/examples/consistency_distillation/train_lcm_distill_lora_sdxl.py#L864C1-L865C64. Somewhat surprisingly, length of `params_to_optimize` (when packed as a list) is zero. This is not expected, no? However, `params_to_optimize_named` prints all the trainable params correctly. 
* I did the same after disabling adapters too: https://github.com/huggingface/diffusers/blob/lcm-lora-sdxl-datasets/examples/consistency_distillation/train_lcm_distill_lora_sdxl.py#L1174C1-L1178C109. It prints expected stuff. No trainable params. 
* Then I come [down](https://github.com/huggingface/diffusers/blob/8fecdda23cefc30053a511db597b9c017b8fcaa1/examples/consistency_distillation/train_lcm_distill_lora_sdxl.py#L1226C1-L1230C108) after enabling the adapters. `print(len(list(params_to_optimize)), len(list(params_to_optimize_after_enable)))` gives `0 1576`, respectively. I think `1576` is expected and should have been the case for `params_to_optimize` as well.

sayakpaul commented:
> Do you have a reference for using this trick for training? I wonder if there could be some kind of flaw in that approach which could explain the findings.

This is taken from https://huggingface.co/blog/stackllama. It's implemented in TRL too:
https://github.com/huggingface/trl/blob/baa8f09cb35057c03b33d898f90c7e8ff958ed9b/trl/trainer/ppo_trainer.py#L468

@lvwerra could you comment more on the above with respect to:

> Instead of initializing two UNets (teacher and student), it leverages the disable_adapters() and enable_adapters() functions. In this case, the teacher is without any adapters and the student is with LoRA. We only update the LoRA params in the student.

sayakpaul commented:
Need to propagate the changes from https://github.com/huggingface/diffusers/pull/6145 once it's merged.

sayakpaul commented:
@dg845 as well if you want to give this a look :-)

dg845 commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1368 @@
+
+
+# From LCMScheduler.get_scalings_for_boundary_condition_discrete
+def scalings_for_boundary_conditions(timestep, sigma_data=0.5, timestep_scaling=10.0):
+    c_skip = sigma_data**2 / ((timestep / 0.1) ** 2 + sigma_data**2)
+    c_out = (timestep / 0.1) / ((timestep / 0.1) ** 2 + sigma_data**2) ** 0.5
+    return c_skip, c_out
```
We could make this a little more general by using the `timestep_scaling` argument (and perhaps expose this as an argument):

```suggestion
def scalings_for_boundary_conditions(timestep, sigma_data=0.5, timestep_scaling=10.0):
    scaled_timestep = timestep_scaling * timestep
    c_skip = sigma_data**2 / (scaled_timestep**2 + sigma_data**2)
    c_out = scaled_timestep / (scaled_timestep**2 + sigma_data**2) ** 0.5
    return c_skip, c_out
```

(The current function is the same as in `examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py` so if we make this change we should probably also propagate it to the WebDataset scripts.)

sayakpaul replied:
Would prefer reflecting that after it's changed in WDS so that it's easier to track.

dg845 commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1368 @@
+
+    # 9. Add LoRA to the student U-Net, only the LoRA projection matrix will be updated by the optimizer.
+    lora_config = LoraConfig(
+        r=args.lora_rank,
+        lora_alpha=args.lora_rank,
```
Would it make sense to allow `lora_alpha` to be set independently of the `r`/`lora_rank`, to allow the LoRA layer scaling to be controlled?

sayakpaul replied:
Maybe in a future PR. Since this PR is almost just a copy-paste of the WDS version.

dg845 commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1368 @@
+                # predicted noise eps_0 and predicted original sample x_0, then run the ODE solver using these
+                # estimates to predict the data point in the augmented PF-ODE trajectory corresponding to the next ODE
+                # solver timestep.
+                unet.disable_adapters()
```
I think it would be a little more clear if a comment was added here to indicate that when the LoRA adapter is disabled, `unet` is exactly the teacher model U-Net, so we can get the teacher outputs from it.

patil-suraj replied:
+1

sayakpaul replied:
Done in 4c689b29.

dg845 commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1368 @@
+                    # Note that the DDIM step depends on both the predicted x_0 and source noise eps_0.
+                    x_prev = solver.ddim_step(pred_x0, pred_noise, index).to(unet.dtype)
+
+                # re-enable unet adapters
+                unet.enable_adapters()
```
Following https://github.com/huggingface/diffusers/pull/5778/files#r1434412096, it might also make sense to modify the comment here to indicate that re-enabling the LoRA adapter turns `unet` back into the student model.

patil-suraj replied:
+1

sayakpaul replied:
Done in 4c689b29.

sayakpaul commented:
@patil-suraj @dg845 ran a recent experiment with the following:

```bash
export MODEL_NAME="stabilityai/stable-diffusion-xl-base-1.0"
export OUTPUT_DIR="lora-lcm-sdxl-new"
export DATASET_NAME="lambdalabs/pokemon-blip-captions"
export VAE_PATH="madebyollin/sdxl-vae-fp16-fix"

CUDA_VISIBLE_DEVICES=1 accelerate launch train_lcm_distill_lora_sdxl.py \
  --pretrained_teacher_model=${MODEL_NAME}  \
  --pretrained_vae_model_name_or_path=${VAE_PATH} \
  --output_dir="lora-lcm-sdxl-new" \
  --mixed_precision="fp16" \
  --dataset_name=$DATASET_NAME \
  --resolution=1024 \
  --train_batch_size=24 \
  --gradient_accumulation_steps=1 \
  --gradient_checkpointing \
  --use_8bit_adam \
  --lora_rank=64 \
  --learning_rate=1e-4 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=10000 \
  --checkpointing_steps=3000 \
  --validation_steps=50 \
  --seed="0" \
  --report_to="wandb" \
  --push_to_hub
```

WandB: https://wandb.ai/sayakpaul/text2image-fine-tune/runs/tv3zw00t
Weights: https://huggingface.co/sayakpaul/pokemons-lora-lcm-sdxl

Given it's only 10k steps and I haven't ablated the hyperparameters, I'd say it's still good enough. WDYT? I'd be keen on merging this soon as it reduces the memory requirements significantly by following good PEFT practices. 

WDYT?

patil-suraj commented:
Sounds good! Usually in my experiment 2-3k steps were enough, 10k results in overfitting with large BS.

patil-suraj commented on examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:
```diff
@@ -0,0 +1,1368 @@
+    return out.reshape(b, *((1,) * (len(x_shape) - 1)))
+
+
+def tokenize_prompt(tokenizer, prompt):
```
Is this used ?

sayakpaul replied:
Removed in c24626ae.

sayakpaul commented:
@patil-suraj I added the following for the reference training command when training on a very small dataset like Pokemons:

```bash
export MODEL_NAME="stabilityai/stable-diffusion-xl-base-1.0"
export DATASET_NAME="lambdalabs/pokemon-blip-captions"
export VAE_PATH="madebyollin/sdxl-vae-fp16-fix"

accelerate launch train_lcm_distill_lora_sdxl.py \
  --pretrained_teacher_model=${MODEL_NAME}  \
  --pretrained_vae_model_name_or_path=${VAE_PATH} \
  --output_dir="pokemons-lora-lcm-sdxl" \
  --mixed_precision="fp16" \
  --dataset_name=$DATASET_NAME \
  --resolution=1024 \
  --train_batch_size=24 \
  --gradient_accumulation_steps=1 \
  --gradient_checkpointing \
  --use_8bit_adam \
  --lora_rank=64 \
  --learning_rate=1e-4 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=3000 \
  --checkpointing_steps=500 \
  --validation_steps=50 \
  --seed="0" \
  --report_to="wandb" \
  --push_to_hub
```

The training dynamics for this aren't clear but https://wandb.ai/sayakpaul/text2image-fine-tune/runs/tv3zw00t definitely shows progress IMO.

sayakpaul commented:
Will merge after the CI is green.
